pkg/controllers/node/hostendpoints.go

```diff
@@ -0,0 +1,296 @@
+	log "github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// deleteAutoHostendpointsWithoutNodes deletes auto hostendpoints that either
+// reference a Calico node that doesn't exist, or, that remain after
+// autoHostEndpoints has been disabled.
+func (c *NodeController) deleteAutoHostendpointsWithoutNodes(heps map[string]api.HostEndpoint) error {
+	for _, hep := range heps {
+		_, hepNodeExists := c.nodeCache[hep.Spec.Node]
+
+		if !hepNodeExists || c.config.AutoHostEndpoints != "enabled" {
+			err := c.deleteHostendpoint(hep.Name)
+			if err != nil {
+				log.WithError(err).Warnf("failed to delete hostendpoint %q", hep.Name)
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// createUpdateAutohostendpoints creates or updates all auto hostendpoints.
+func (c *NodeController) createUpdateAutohostendpoints() error {
+	for _, node := range c.nodeCache {
+		err := c.syncAutoHostendpoint(node)
+		if err != nil {
+			log.WithError(err).Warnf("failed to sync hostendpoint for node %q", node.Name)
+			return err
+		}
+	}
+	return nil
+}
+
+// syncAllAutoHostendpoints ensures that the expected auto hostendpoints exist,
+func (c *NodeController) syncAllAutoHostendpoints() error {
+	for n := 1; n <= 5; n++ {
+		log.Debugf("syncing all hostendpoints. attempt #%v", n)
+		autoHeps, err := c.listAutoHostendpoints()
+		if err != nil {
+			log.WithError(err).Warn("failed to list hostendpoints")
+			time.Sleep(retrySleepTime)
+			continue
+		}
+
+		// Delete any dangling auto hostendpoints
+		if err := c.deleteAutoHostendpointsWithoutNodes(autoHeps); err != nil {
+			log.WithError(err).Warn("failed to delete dangling hostendpoints")
+			time.Sleep(retrySleepTime)
+			continue
+		}
+
+		// For every Calico node in our cache, create/update the auto hostendpoint
+		// for it.
+		if c.config.AutoHostEndpoints == "enabled" {
+			if err := c.createUpdateAutohostendpoints(); err != nil {
+				log.WithError(err).Warn("failed to sync hostendpoint for nodes")
+				time.Sleep(retrySleepTime)
+				continue
+			}
+		}
+
+		log.Info("successfully synced all hostendpoints")
+		return nil
+	}
+	return fmt.Errorf("too many retries when syncing all hostendpoints")
+}
+
+// syncAutoHostendpoint syncs the auto hostendpoint for the given node.
+func (c *NodeController) syncAutoHostendpoint(node *api.Node) error {
+	hepName := c.generateAutoHostendpointName(node.Name)
+	log.Debugf("syncing hostendpoint %q from node %+v", hepName, node)
+
+	// Try getting the host endpoint.
+	expectedHep := c.generateAutoHostendpointFromNode(node)
+	currentHep, err := c.calicoClient.HostEndpoints().Get(c.ctx, hepName, options.GetOptions{})
+	if err != nil {
+		switch err.(type) {
+		case errors.ErrorResourceDoesNotExist:
+			log.Infof("host endpoint %q doesn't exist, creating it", node.Name)
+			if _, err := c.createAutoHostendpoint(node); err != nil {
+				log.WithError(err).Warnf("failed to create host endpoint %q, retrying", node.Name)
```

Comment:
These logs probably shouldn't say retrying, since there is no longer a retry loop in this function and it is handled by the calling code.

You might just want to get rid of these altogether and us the calling code to log out, since the calling code knows if this is an Error level or a Warning level.

Reply:
7da10da